src/test/groovy/com/calgaryscientific/gradle/VeracodeTaskTest.groovy

```diff
@@ -192,16 +194,69 @@ class VeracodeTaskTest extends Specification {
 
         when:
         project.plugins.apply('com.calgaryscientific.gradle.veracode')
-        project.ext.veracodeSetup = vs
+        project.veracodeSetup.filesToUpload = vs.filesToUpload
 
         then:
         VeracodeSetup vsRead = project.findProperty("veracodeSetup") as VeracodeSetup
         Set<File> expected = [buildFile] as Set
         assert vsRead.filesToUpload == expected
-        def task = project.tasks.getByName("veracodeUploadFile")
+        VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
+        assert task.veracodeSetup.filesToUpload == vsRead.filesToUpload
         assert task.getFileSet() == expected
         def _ = vsRead.filesToUpload.add(buildFile)
         assert vsRead.filesToUpload == [buildFile, buildFile] as Set
     }
 
+    def 'Test VeracodeUploadFile Task'() {
+        given:
+        VeracodeAPI veracodeAPIMock = Mock(VeracodeAPI, constructorArgs: ["", "", null, null])
+        String xmlStr = '''<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<filelist xmlns="something" xmlns:xsi="something" filelist_version="1.1">
+    <file file_id="1" file_md5="d98b6f5ccfce3799e9b60b5d78cc1" file_name="file1" file_status="Uploaded"/>
+    <file file_id="2" file_md5="68a7d8468ca51bc46d5b72d485022" file_name="file2" file_status="Uploaded"/>
+    <file file_id="3" file_md5="2459464ff4bf78dd6f09695069b52" file_name="file3" file_status="Uploaded"/>
+</filelist>
+'''
+        Project project = new ProjectBuilder().build()
+        project.plugins.apply('com.calgaryscientific.gradle.veracode')
+        VeracodeSetup vs = new VeracodeSetup()
+        vs.filesToUpload = project.fileTree(dir: testProjectDir.root, include: '**/*').getFiles()
+        project.veracodeSetup.filesToUpload = vs.filesToUpload
+        VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
+        task.veracodeAPI = veracodeAPIMock
+
+        when:
+        task.run()
+
+        then:
+        1 * veracodeAPIMock.uploadFile(_, _) >> {
+            return xmlStr
+        }
+    }
+
+    def 'Test VeracodeUploadFile Task failure'() {
+        given:
+        VeracodeAPI veracodeAPIMock = Mock(VeracodeAPI, constructorArgs: ["", "", null, null])
+        String xmlStr = '''<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<error>Could not upload file</error>
+'''
+        Project project = new ProjectBuilder().build()
```

Comment:
I think you can remove the duplication by moving the common setup code to setup function
such as: 
```
Project project = new ProjevtBuilder().build();
project.plugins.apply('com.calgaryscientific.gradle.veracode')
VeracodeSetup vs = new VeracodeSetup()
vs.filesToUpload = project.fileTree(dir: testProjectDir.root, include: '**/*').getFiles()
project.veracodeSetup.filesToUpload = vs.filesToUpload
VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
```

Reply:
It does seem like duplication but the common setup is only valid for those two tests, the rest of the tests will not share that. But I still can do that.
I thought about that for a while and initially decided against it because it is not a common setup for all tests.

Only this part is common to the upcoming tests:

````
Project project = new ProjevtBuilder().build();
project.plugins.apply('com.calgaryscientific.gradle.veracode')
````

The rest is specific for this one task:

````
VeracodeSetup vs = new VeracodeSetup()
vs.filesToUpload = project.fileTree(dir: testProjectDir.root, include: '**/*').getFiles()
project.veracodeSetup.filesToUpload = vs.filesToUpload
VeracodeUploadFileTask task = project.tasks.getByName("veracodeUploadFile") as VeracodeUploadFileTask
````

But I can do a setupUploadTaskProject function that applies to only these two tasks.

Reply:
- sounds like you can extract the minimal common code to the setup
- the bit that is not common to everything, you could extract that to a function so other tests can reuse that utilities function, however, no other tests is using it for now, so you don't need to extract that to a function at the moment.